From 36a4423d82291f6e01ee26a6f5bf0b5d0d75d07e Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 17 Jul 2015 14:31:49 +0800
Subject: [PATCH] [PATCH] dhcp: add exec script function

Make dhclient to run /etc/dhcp/dhclient.d/*.sh, and get the configuration
for ntp, nis or other programes

These codes are from redhat rpm packages.

Upstream-Status: Inappropriate [embedded specific]

Signed-off-by: Li Wang <li.wang@windriver.com>
Signed-off-by: Roy Li <rongqing.li@windriver.com>
Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 client/scripts/linux | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/client/scripts/linux b/client/scripts/linux
index e9ea379..716d7e8 100755
--- a/client/scripts/linux
+++ b/client/scripts/linux
@@ -31,6 +31,8 @@
 # overwirte this line to use a fake ip-echo tool. It's also convenient
 # if your system holds ip tool in a non-standard location.
 ip=/sbin/ip
+ETCDIR="/etc/dhcp"
+export SAVEDIR=/var/lib/dhcp
 
 # update /etc/resolv.conf based on received values
 # This updated version mostly follows Debian script by Andrew Pollock et al.
@@ -182,6 +184,20 @@ exit_with_hooks() {
     exit $exit_status
 }
 
+execute_client_side_configuration_scripts() {
+# execute any additional client side configuration scripts we have
+    if [ "${1}" == "config" ] || [ "${1}" == "restore" ]; then
+        for f in ${ETCDIR}/dhclient.d/*.sh ; do
+            if [ -x "${f}" ]; then
+                subsystem="${f%.sh}"
+                subsystem="${subsystem##*/}"
+                . "${f}"
+                "${subsystem}_${1}"
+            fi
+        done
+    fi
+}
+
 # This function was largely borrowed from dhclient-script that
 # ships with Centos, authored by Jiri Popelka and David Cantrell
 # of Redhat. Thanks guys.
@@ -325,10 +341,12 @@ case "$reason" in
 
         # update /etc/resolv.conf
         make_resolv_conf
+        execute_client_side_configuration_scripts "config"
 
         ;;
 
     EXPIRE|FAIL|RELEASE|STOP)
+        execute_client_side_configuration_scripts "restore"
         if [ -n "$alias_ip_address" ]; then
             # flush alias IP
             ${ip} -4 addr flush dev ${interface} label ${interface}:0
@@ -464,6 +482,7 @@ case "$reason" in
            [ "${new_dhcp6_domain_search}" != "${old_dhcp6_domain_search}" ]; then
             make_resolv_conf
         fi
+        execute_client_side_configuration_scripts "config"
 
         ;;
 
@@ -476,6 +495,7 @@ case "$reason" in
         ${ip} -6 addr change ${cur_ip6_address}/${cur_ip6_prefixlen} \
             dev ${interface} scope global preferred_lft 0
 
+        execute_client_side_configuration_scripts "config"
         ;;
 
     EXPIRE6|RELEASE6|STOP6)
@@ -487,6 +507,7 @@ case "$reason" in
         ${ip} -6 addr del ${old_ip6_address}/${old_ip6_prefixlen} \
             dev ${interface}
 
+        execute_client_side_configuration_scripts "restore"
         ;;
 esac
 
